lib/full-site-editing/block-templates.php

```diff
@@ -241,31 +241,35 @@ function _gutenberg_build_template_result_from_file( $template_file, $template_t
 /**
  * Build a unified template object based a post Object.
  *
- * @param WP_Post $post Template post.
+ * @param WP_Post $post          Template post.
+ * @param string  $template_type wp_template or wp_template_part.
  *
  * @return WP_Block_Template|WP_Error Template.
  */
-function _gutenberg_build_template_result_from_post( $post ) {
-	$terms = get_the_terms( $post, 'wp_theme' );
-
-	if ( is_wp_error( $terms ) ) {
-		return $terms;
+function _gutenberg_build_template_result_from_post( $post, $template_type = 'wp_template' ) {
+	if ( $template_type !== $post->post_type ) {
+		return new WP_Error( 'template_wrong_post_type', __( 'An invalid post was provided for this template.', 'gutenberg' ) );
```

Comment:
I'll be leaving another round of review comments, partly to inform the `wordpress-develop` minimal backport, see https://github.com/WordPress/gutenberg/pull/31971#issuecomment-855943266.

---

I'd drop the `$template_type` argument. I don't really see much value in it TBH; let's just infer the template type from `$post->post_type`, and make sure that it's found in a list of allowed types:

```suggestion
function _gutenberg_build_template_result_from_post( $post ) {
	if ( ! in_array( $post->post_type, array( 'wp_template', 'wp_template_part', true ) ) {
		return new WP_Error( 'template_wrong_post_type', __( 'An invalid post was provided for this template.', 'gutenberg' ) );
```

Reply:
Amended.